cpp/src/fil/infer.cu

```diff
@@ -452,6 +457,87 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES> {
   }
 };
 
+template <int NITEMS>
+struct tree_aggregator_t<NITEMS, VECTOR_LEAF> {
+  float* probability_sum;
+  int num_classes;
+  val_t* vector_leaf;
+
+  static size_t smem_finalize_footprint(size_t data_row_size, int num_classes,
+                                        bool predict_proba) {
+    return 0;
+  }
+  static size_t smem_accumulate_footprint(int num_classes) {
+    return sizeof(float) * num_classes * NITEMS;
+  }
+
+  __device__ __forceinline__ tree_aggregator_t(predict_params params,
+                                               void* accumulate_workspace,
+                                               void* finalize_workspace,
+                                               val_t* vector_leaf)
+    : num_classes(params.num_classes),
+      probability_sum((float*)accumulate_workspace),
+      vector_leaf(vector_leaf) {
+    for (int c = threadIdx.x; c < num_classes; c += FIL_TPB * NITEMS)
+#pragma unroll
+      for (int item = 0; item < NITEMS; ++item)
+        probability_sum[c * NITEMS + item] = 0.0f;
+    // __syncthreads() is called in infer_k
+  }
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, int> single_tree_prediction, int tree, int num_rows) {
+#pragma unroll
+    for (int item = 0; item < NITEMS; ++item) {
+      for (int k = 0; k < num_classes; ++k) {
+        raft::myAtomicAdd(probability_sum + k * NITEMS + item,
+                          vector_leaf[single_tree_prediction[item] + k].f);
+      }
+    }
+  }
+  // class probabilities or regression. for regression, num_classes
+  // is just the number of outputs for each data instance
+  __device__ __forceinline__ void finalize_multiple_outputs(float* out,
+                                                            int num_rows,
+                                                            int num_trees) {
+    __syncthreads();
+    for (int c = threadIdx.x; c < num_classes; c += blockDim.x) {
+#pragma unroll
+      for (int row = 0; row < num_rows; ++row) {
+        out[row * num_classes + c] = probability_sum[c * NITEMS + row];
+      }
+    }
+  }
+  // using this when predicting a single class label, as opposed to sparse class vector
+  // or class probabilities or regression
+  __device__ __forceinline__ void finalize_class_label(float* out,
+                                                       int num_rows) {
+    __syncthreads();
+    int item = threadIdx.x;
+    int row = item;
+    if (item < NITEMS && row < num_rows) {
```

Comment:
same request to reorder loops and transpose array. We have an argmax reduce template that you can just use, with better parallelism. It will mean reducing regardless of real row count, and just discard the extra garbage before writing to global memory in the last block.
I didn't get around to improve CATEGORICAL_LEAF the same way, but I will eventually.